cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/detector/BaseGoalViolationDetector.java

```diff
@@ -0,0 +1,228 @@
+        LOG.debug("Skipping goal violation detection because the model generation hasn't changed. Current model generation {}",
+                _kafkaCruiseControl.loadMonitor().clusterModelGeneration());
+      }
+      return AnomalyDetectionStatus.SKIP_MODEL_GENERATION_NOT_CHANGED;
+    }
+
+    AnomalyDetectionStatus detectionStatus = getAnomalyDetectionStatus(_kafkaCruiseControl, true, true);
+    if (detectionStatus == AnomalyDetectionStatus.SKIP_HAS_OFFLINE_REPLICAS) {
+      setBalancednessWithOfflineReplicas();
+    } else if (detectionStatus == AnomalyDetectionStatus.SKIP_EXECUTOR_NOT_READY) {
+      // An ongoing execution might indicate a cluster expansion/shrinking. Hence, the detector avoids reporting a stale provision status.
+      _provisionResponse = new ProvisionResponse(ProvisionStatus.UNDECIDED);
+      // An ongoing execution may modify the replication factor of partitions; hence, the detector avoids reporting potential false positives.
+      _hasPartitionsWithRFGreaterThanNumRacks = false;
+    }
+
+    return detectionStatus;
+  }
+
+  /**
+   * @param clusterModel The state of the cluster.
+   * @return {@code true} to skip goal violation detection due to offline replicas in the cluster model.
+   */
+  protected boolean skipDueToOfflineReplicas(ClusterModel clusterModel) {
+    if (!clusterModel.deadBrokers().isEmpty()) {
+      LOG.info("Skipping goal violation detection due to dead brokers {}, which are reported by broker failure "
+              + "detector, and fixed if its self healing configuration is enabled.", clusterModel.deadBrokers());
+      setBalancednessWithOfflineReplicas();
+      return true;
+    } else if (!clusterModel.brokersWithBadDisks().isEmpty()) {
+      LOG.info("Skipping goal violation detection due to brokers with bad disks {}, which are reported by disk failure "
+              + "detector, and fixed if its self healing configuration is enabled.", clusterModel.brokersWithBadDisks());
+      setBalancednessWithOfflineReplicas();
+      return true;
+    }
+
+    return false;
+  }
+
+  protected void setBalancednessWithOfflineReplicas() {
+    _balancednessScore = BALANCEDNESS_SCORE_WITH_OFFLINE_REPLICAS;
+    _provisionResponse = new ProvisionResponse(ProvisionStatus.UNDECIDED);
+  }
+
+  protected void refreshBalancednessScore(Map<Boolean, List<String>> violatedGoalsByFixability) {
+    _balancednessScore = MAX_BALANCEDNESS_SCORE;
+    for (List<String> violatedGoals : violatedGoalsByFixability.values()) {
+      violatedGoals.forEach(violatedGoal -> _balancednessScore -= _balancednessCostByGoal.get(violatedGoal));
+    }
+  }
+
+  protected Set<String> excludedTopics(ClusterModel clusterModel) {
+    return Utils.getTopicNamesMatchedWithPattern(_excludedTopics, clusterModel::topics);
+  }
+
+  protected boolean optimizeForGoal(ClusterModel clusterModel,
+                                    Goal goal,
+                                    GoalViolations goalViolations,
```

Comment:
If we change the type to AbstractGoalViolations, maybe we can reuse this method for both GoalViolations and IntraBrokerGoalViolations?